Test attribution status in new page

Malar-Natarajan opened the pull request:
GIVEN: A page URL in the format of {base_url}/books/{book_slug}/pages/{page_slug}
AND: The citation/attribution tab is open
WHEN: navigating via (toc link, next link, previous link)
THEN: the citation/attribution section is not open on the new page

Malar-Natarajan committed:
tests attribution status in next and previous pages

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
verifying next link locators

Malar-Natarajan committed:
next and previous works but with sleep

Malar-Natarajan committed:
playing around toc

Malar-Natarajan committed:
playing around toc

Malar-Natarajan committed:
playing with toc

Malar-Natarajan committed:
toc chapter click works

Malar-Natarajan committed:
page click and chapter click works

Malar-Natarajan committed:
test works but with sleep

Malar-Natarajan committed:
test works good without sleep

Malar-Natarajan committed:
test works good without sleep

Malar-Natarajan committed:
test works

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
some minor changes not affecting the test

Malar-Natarajan committed:
Merge branch 'test-attribution-new-page' of https://github.com/openstax/rex-web into test-attribution-new-page

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -1,20 +1,36 @@
 
+    @property
+    def title(self):
+        return self.find_element(*self._title_locator)
```
titiles apply to all pages, not just content pages, this could go in the super class

Malar-Natarajan replied:
created a helper function for title. I shall add this as super class when we have more pages coming into Rex

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -31,6 +47,24 @@ def sidebar(self):
+        self.offscreen_click(self.next_link)
+        self.wait.until(
+            lambda _: title_before_click != (self.title.get_attribute("innerHTML") or "")
+        )
```
this `wait.util` on the title is neat, but maybe we should move it to a helper function so we don't have to put it in multiple places? and then if we ever update the logic for being able to tell when navigation is done it'll only be in one place?

Malar-Natarajan replied:
created a helper function

TomWoodward commented on pytest-selenium/tests/test_attribution.py:
```diff
@@ -49,3 +50,59 @@ def test_attribution_collapsed_by_default_expands_when_clicked(
+    attribution.click_attribution_link()
+
+    # WHEN: Navigating via Back link
+    content.click_previous_link()
```
i think it might be better to have 3 separate tests for next/prev/toc

Malar-Natarajan replied:
@TomWoodward 
Wouldn't that be too overwhelming to have just two lines of code in one test...I preferred this way because testrail test was designed like this to include all possible navigations in one test

Malar-Natarajan replied:
I removed the TOC part from this test since the TOC structure needs more time for me to put in a neat framework

TomWoodward commented on pytest-selenium/tests/test_attribution.py:
```diff
@@ -49,3 +50,59 @@ def test_attribution_collapsed_by_default_expands_when_clicked(
+
+        # THEN: the citation/attribution section is not open on the new page
+        assert not attribution.is_open
+        attribution.click_attribution_link()
```
why click it at the end?

Malar-Natarajan replied:
@TomWoodward 
No great reason. But my hypothesis is just making sure attribution is clickable after navigating to a new page via TOC. 
I can remove that if it hurts you :-)

TomWoodward commented on pytest-selenium/tests/test_attribution.py:
```diff
@@ -49,3 +50,59 @@ def test_attribution_collapsed_by_default_expands_when_clicked(
+    # WHEN: Navigating via TOC link
+    toolbar = content.toolbar
+    toc = content.sidebar.toc
+    from time import sleep
```
unused?

Malar-Natarajan replied:
import sleep was my bad. I forgot to remove it.
The toolbar assignment is used with the the mobile resolution. I moved under the mobile If clause now.
The toc assignment is used in both desktop & mobile. So stays in the same place

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -78,6 +116,68 @@ def click_toc_toggle_button(self):
+            _toc_chapter_name_locator = (By.XPATH, "(//ol/li/details/summary/div/span)")
+
+            @property
+            def is_chapter_expanded(self):
```
this seems to be unused

Malar-Natarajan replied:
removed

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -78,6 +116,68 @@ def click_toc_toggle_button(self):
+                return self.find_element(*self._chapter_toggle).get_attribute("open")
+
+            @property
+            def number_of_chapters(self):
```
also this

Malar-Natarajan replied:
removed

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -78,6 +116,68 @@ def click_toc_toggle_button(self):
+                ]
+
+            @property
+            def chapter_name(self):
```
and this

Malar-Natarajan replied:
removed

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -78,6 +116,68 @@ def click_toc_toggle_button(self):
+                    ]
+
+                @property
+                def number_of_pages(self):
```
also unused?

Malar-Natarajan replied:
Removed

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -78,6 +116,68 @@ def click_toc_toggle_button(self):
+                return len(self.find_elements(*self._chapter_toggle))
+
+            @property
+            def chapters(self):
```
you have this organized such that the toc has chapters and the chapter have pages but that is not true, the toc can have pages directly and the chapters can have more chapters inside

Malar-Natarajan replied:
TOC is removed from this PR. But I will keep this point in mind when doing TOC

TomWoodward commented on pytest-selenium/regions/content_item.py:
```diff
@@ -0,0 +1,40 @@
+        super().__init__(page, root)
+
+    @property
+    def has_chapter_section(self):
```
unused?

Malar-Natarajan replied:
removed. 
I was playing around the TOC elements and added all these. I will make sure not to add unised attributes/elements in future

TomWoodward commented on pytest-selenium/regions/content_item.py:
```diff
@@ -0,0 +1,40 @@
+        return self.find_element(*self._chapter_section_span_locator)
+
+    @property
+    def chapter_section(self):
```
unused?

Malar-Natarajan replied:
removed

TomWoodward commented on pytest-selenium/regions/content_item.py:
```diff
@@ -0,0 +1,40 @@
+        return self.find_element(*self._title_span_locator)
+
+    @property
+    def title(self):
```
uused?

Malar-Natarajan replied:
removed

Malar-Natarajan committed:
updated based on test_attribution review comment

Malar-Natarajan committed:
removed unused attributes

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
next element click with title verification works as helper function but still code duplication in base.py

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'test-attribution-new-page' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Next link helper function kind of works

Malar-Natarajan committed:
helper functions created for TOC links

Malar-Natarajan committed:
removed a redundant offscreen click method in pages/base.py
Still have the original method as commented out. Can revert if my changes are wrong

Malar-Natarajan committed:
finishing C476304
removed TOC navigation from this test. It will be covered later

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

TomWoodward commented on pytest-selenium/pages/base.py:
```diff
@@ -44,3 +44,15 @@ def offscreen_click(self, element):
         element.send_keys(Keys.ENTER)
         return element
+
+    def offscreen_click_and_wait_for_new_title_to_load(self, element):
```
name binds to implementation, how about

```suggestion
    def click_and_wait_for_load(self, element):
```

Malar-Natarajan replied:
Changed as mentioned

TomWoodward commented on pytest-selenium/pages/base.py:
```diff
@@ -44,3 +44,15 @@ def offscreen_click(self, element):
+        Clicks the given element, even if it is offscreen, by sending the ENTER key.
+        Returns after loading the last element (title) of the page).
+        """
+        title_before_click = self.title_before_click
```
this function is using other functions defined in the subclass, they need to be moved here

TomWoodward replied:
@Malar-Natarajan this still needs to be fixed

Malar-Natarajan replied:
@TomWoodward did you mean that I commented the `offscreen_click()` subclass but forgot to remove it or did you mean some other changes?

Malar-Natarajan replied:
Updated in the latest commit

TomWoodward commented on pytest-selenium/regions/base.py:
```diff
@@ -16,4 +18,8 @@ def offscreen_click(self, element=None):
+        # We actually navigate using the ENTER key because scrolling the page can be flaky
+        # https://stackoverflow.com/a/39918249
+        # return self.page.offscreen_click(element or self.root)
+        element.send_keys(Keys.ENTER)
```
i wonder if this'll fix the tests in edge 🤔

TomWoodward commented on pytest-selenium/regions/content_item.py:
```diff
@@ -0,0 +1,28 @@
+        root = parent_root.find_element(
+            By.XPATH, self._root_locator_template.format(index=index + 1)
+        )
+        self.parent_root = parent_root
```
is this used for something?

Malar-Natarajan replied:
Removed with lates commit. This file is not needed now since we removed the TOC part from this PR

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
Deleting content_item.py 

This file was needed for the TOC changes. Since the TOC part is removed from this PR, I am deleting this file

Malar-Natarajan committed:
updated review comments

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

m1yag1 committed:
Merge branch 'master' of github.com:openstax/rex-web into test-attribution-new-page

m1yag1 committed:
refactored click_and_wait to use a reference object and renamed the method

m1yag1 committed:
added offscreen_click back to the region

m1yag1 committed:
refactored content page to use updated click and check reference object on base page

m1yag1 committed:
update docstring on click_and_check helper method

TomWoodward commented on pytest-selenium/pages/base.py:
```diff
@@ -44,3 +44,23 @@ def offscreen_click(self, element):
         element.send_keys(Keys.ENTER)
         return element
+
+    def click_and_check_reference_object_has_changed(self, click_element, ref_object, ref_property):
```
checking to make sure the page title or chapter title have been changed are nifty but not completely accurate ways to make sure the page has loaded, i liked the previous implementation of this method because it abstracted the exact method of telling when the page was loaded, and if we implement a better way in the future we'd only have to do it in one place.

m1yag1 replied:
i don't see how it abstracted anything away as it was hardcoded to always use the title. It was also checking the title on the parent class which didn't have a title which was also confusing to me. in this case here, any element that is clicked that also causes a side effect can be checked. 

I'm in full agreement that there should be something better for detecting when a page is fully loaded. I wanted to suggest initially that maybe on these navigation transitions maybe the `rex-loaded` attribute should be removed and added again when the page is loaded. @TomWoodward

TomWoodward replied:
yea i agree that a better solution is possible and we should do it, but it doesn't have to happen right now and we can just change it in this method when we get around to it

m1yag1 replied:
i agree. i've reverted my changes.

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -50,6 +78,10 @@ class BookBanner(Region):
         _chapter_section_locator = (By.CSS_SELECTOR, "div > h1 > span.os-number")
 
+        @property
+        def book_title_element(self):
```
didn't mike add a bookbanner region in #188 ? can't we use that here?

m1yag1 replied:
oh i think that's left over from a different thing i was trying. ooops.

m1yag1 committed:
Revert "update docstring on click_and_check helper method"

This reverts commit adf82be7cb08cd6a55f024bcebaf61de0d3a3986.

TomWoodward commented on pytest-selenium/tests/test_attribution.py:
```diff
@@ -49,3 +50,31 @@ def test_attribution_collapsed_by_default_expands_when_clicked(
+    # THEN: The citation/attribution section is not open on the new page
+    assert not attribution.is_open
+
+    attribution.click_attribution_link()
```
there is no assertion after this, so its not testing anything right? why is it here?

Malar-Natarajan replied:
my bad..removed it

Malar-Natarajan committed:
updated the title to be in base.py

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page
